### DISCUSSION TITLE
Wysyłanie zapytań do API KeyVault (albo innych MS) przez API Management

### DISCUSSION THREAD

**Discussion by krukowskid** [AzureCommunityPL/group#240] [Pomoc] [answered]
Próbuję skonfigurować API Management tak, aby można było wywołać funkcję decrypt na KeyVault i dostać w odpowiedzi wynik.

Mój plan:
Wysyłam zapytania do API Management, którego tożsamość zarządzana ma możliwość wykonania decrypt przy użyciu klucza w KeyVault. KeyVault zwraca wynik do API Management, a API Management przekazuje go mi. 

Od strony KV konfiguracja jest dla mnie jasna, ale pokonała mnie konfiguracja reguł API Management.

Z chęcią przyjmę każde info na temat tego jak to powinno wyglądąć :)

**Comment by mgrabarz** [comment, marked as answer]
W polityce potrzebujesz wykonać następjące kroki. Niestety nie mam pod ręką gotowca:

1. Wsadzenie zaszyfrowanej wartości do zmiennej
2. Wysłać żadanie do KV:
```xml
<send-request mode="new" timeout="20" ignore-error="false" response-variable-name="decrypted">
    <set-url>{vaultBaseUrl}/keys/{key-name}/{key-version}/decrypt?api-version=7.1</set-url>
    <set-method>POST</set-method>
    <authentication-managed-identity resource="https://vault.azure.net"/>
    <set-body>{
        "alg": "RSA-OAEP",
        "value": "wartosc_do_odszyfrowania_ze_zmiennej"
      }</set-body>
</send-request>

```
3. Wyciągnąć do zmiennej odszyfrowaną wartość z "decrypted"

Przy czym Set-body musi być nie tak jak powyżej, tylko jak w samplu: https://docs.microsoft.com/en-us/azure/api-management/api-management-sample-send-request.
Oczywiście Twój APIM musi mieć MSI włączone (nie wiem czy user-assigned juz wspiera ten scenariusz), MSI musi mieć uprawnienia w KV, APIM musi miec wjazd do KV po publicznym IP, lub poprzez Service Endopint - Private Endpoint nie jest jeszcze wspierany.

**Reply by krukowskid** [reply]
Dzięki, trochę mi to rozjaśniło temat. 
Widzę, że problem mam chyba nie z odpytaniem KV, bo patrząc na logi z KeyVault wygląda na to, że API się do niego dostaje, a zwracanie wartości - jak wysyłam zapytanie (testuję z poziomu PowerShell) to dostaję pustą odpowiedź.

**Reply by krukowskid** [reply]
Udało mi się wyprodukować coś takiego. Działa, ale czy nie zrobiłem tu jakiegoś druta? 
```
<policies>
    <inbound>
        <base />
        <send-request mode="new" response-variable-name="vault-key" timeout="20" ignore-error="false">
            <set-url>https://{{keyvault-name}}.vault.azure.net/keys/{{keyvault-id}}/{{keyvault-keyversion}}/decrypt?api-version=7.1</set-url>
            <set-method>POST</set-method>
            <set-header name="Content-Type" exists-action="override">
                <value>application/json</value>
            </set-header>
            <set-body>@{ 
                JObject inBody = context.Request.Body.As<JObject>(); 
                return inBody.ToString(); 
            }</set-body>
            <authentication-managed-identity resource="https://vault.azure.net" />
        </send-request>
    </inbound>
    <backend>
        <return-response response-variable-name="existing context variable">
            <set-status code="200" />
            <set-body>@{
                JObject identity = ((IResponse)context.Variables["vault-key"]).Body.As<JObject>();
                return identity.SelectToken("value").ToString();
            }</set-body>
        </return-response>
    </backend>
    <outbound>
        <base />
    </outbound>
    <on-error>
        <base />
    </on-error>
</policies>
```

**Reply by mgrabarz** [reply]
Wygląda jak API przelotka do KV. 

Użytkownicy zamiast strzelać bezpośrednio, strzelają na decrypt w APIM z pominięciem Access Policies na KV - nie znam Twojego tematu ale sam design wydaje mi się nieco niekoszerny :P

**Reply by krukowskid** [reply]
Tak, może się to wydawać bez sensu, bo nie ma w tym kodzie kluczowego elementu - limit zapytań i reguł firewall, które pominąłem bo są łatwe w konfiguracji. Backend (nie user) ma strzelać do KeyVault, ale dostęp do KeyVault ma być ograniczony do IP backendu (można zrobić po stronie KV) i musi być limit na decrypt - w tym celu jest APIM, a przy okazji zwraca sam odszyfrowany ciąg bez nazwy KV. Mam nadzieję, że teraz to ma więcej sensu, ale jeżeli masz jakieś rady/pomysły to chętnie je przeczytam i rozliczę się odpowiednio jak będą kiedyś spotkania grupy w trybie offline :)

**Reply by kamil-mrzyglod** [reply]
Tutaj bardziej Markowi chodziło, że trochę to wypacza koncepcję zabezpieczenia KV. Normalnie używasz Access Policies aby limitować kto i na jakich zasadach może wykonać odpowiednią operację na poziomie KV. W przypadku jak robisz przelotkę APIM -> KV to de facto dajesz dowolnemu klientowi(który ma dostęp do APIMa) dostęp do KV na poziomie operacji, którą sobie zaplanowałeś. Co innego jakbyś próbował zautoryzować się on-behalf-of za pomocą principala, który woła APIMa - wtedy mógłbyś wykorzystać limit na decrypt jak i Access Policies.

**Reply by krukowskid** [reply]
@kamil-mrzyglod principal jest pomiędzy APIM <-> KV, które są w innym tenancie niż backend, więc uznałem to za najsensowniejsze rozwiązanie.

**Reply by kamil-mrzyglod** [reply]
@krukowskid Taki setup na pewno komplikuje sprawę, wiadomo, mógłbyś mieć coś, co tłumaczy identity klienta z jednego tenanta na drugiego. Ostatecznie Twoje rozwiązanie pewnie nie jest złe, da się też zrobić je po prostu lepiej :)

**Reply by krukowskid** [reply]
@kamil-mrzyglod jeśli się da to chętnie zobaczę jakieś wskazówki. Powyższe rozwiązanie to na razie POC. Chętnie przetestuję inne  rozwiązania i się czegoś nauczę. Ta kombinacja o której piszę przyszła mi jako jedyna do głowy jeżeli chodzi o wykorzystanie 100% Azure. Backend to web app na Azure w jednym tenancie, KV jest w drugim zupełnie niepowiązanym tenancie. KV musi być ograniczony do zapytań tylko z backendu i mieć ograniczoną ilość wywołań API. Rozwiązanie musi być przede wszystkim bezpieczne, z mniejszym naciskiem na dostępność/cenę/łatwość wdrożenia i utrzymania

**Reply by kaluzaaa** [reply]
czemu kv jest w innym tenancie i jakie masz ku temu powód?

**Reply by krukowskid** [reply]
@kaluzaaa chodzi o zupełne oddzielenie ról, dostępu, audyt. Łatwiej jest to upilnować w oddzielnym tenancie. W przypadku przejęcia kontroli nad tenantem A, można łatwo i szybko odciąć dostęp do KV w tenancie B.

**Reply by kaluzaaa** [reply]
> chodzi o zupełne oddzielenie ról, dostępu, audyt. Łatwiej jest to upilnować w oddzielnym tenancie.

To teraz uwierz w 100% - to wielki błąd, serio ;-)

**Reply by krukowskid** [reply]
Zedytowałem lekko odpowiedź. Oczywiście Ci wierzę i myślę, że odpowiedź brzmi "to zależy". Jeżeli to będzie drugi tenant żyjący swoim radosnym życiem to będzie trudno to upilnować. Mam na myśli skonfigurowaną raz przestrzeń i nadmiernie chronionym dostępem do konta z możliwością logowania do tenantu. W tym scenariuszu moim zdaniem wspólny tenant < oddzielny tenant

**Reply by kaluzaaa** [reply]
Są pewne scenariusze gdzie oddzielny sens ma inny tenant, ale to wtedy jest cała aplikacja tam wsadzana, np. jakiś saas.

W innych przypadkach to kocham to podejście bo wiadomo, że kiedyś będzie trzeba posprzątać porządnie po obsecurity ;-)

**Reply by krukowskid** [reply]
Dziękuję @mgrabarz @kamil-mrzyglod @kaluzaaa za dawkę wiedzy. Tytułowy temat wyczerpany, a zagadnienia powiązane, na które zwróciliście mi uwagę przeanalizuję i spróbuje wybrać to co będzie najlepsze. Każdemu wiszę przynajmniej po piwku ;)
